### DISCUSSION TITLE
How to use 2 midi controllers with each other?

### DISCUSSION THREAD

**Discussion by BrahRah** [helgoboss/helgobox#720] [Q&A] [answered]
Using two devices to control a vst with "control input" set to fx doesn't seem to work. I got it working by using 2 instances of realearn one "control input" is set to device 1 the other to device 2. That works to control one vst with two controllers. When I tried to use device 2 to change presets on device 1 it doesn't work. Device 2 can only control its own parameters not the parameters for the 2nd instance of realearn. Is it possible to change presets for other controllers or am I out of luck?

**Comment by helgoboss** [comment, marked as answer]
Hi. What do you mean by "changing presets for other controllers"? ReaLearn has a target to change presets of e.g. VSTs in REAPER. There's no dedicated target to change presets of hardware devices. Usually, hardware devices listen to program change messages and in turn change presets. Is that what you are looking for? In this case, use the "MIDI: Send message" target.

**Comment by BrahRah** [comment]
I'm not too familiar with the terminology I'll make screenshots.

I meant mapping group not preset sorry for the confusion.
Let my try to explain it better. My controllers don't have enough knobs and buttons and the studiologics bank buttons don't work. So I tried to use my mpd218 drumpad to change the studiologics mapping groups via the "when bank is selected" method. But that seem to only work when the same controller is used not when using one controller to change the mapping group of another controller with it's own main preset and realearn fx instance.

I also noticed that when I try to use multiple controllers on one vst like this:
![Screenshot 2022-10-22 13:55:38](https://user-images.githubusercontent.com/40721994/197337955-c334ca6f-daeb-4c42-a272-69156da34934.png)


Certain knobs and buttons are used at the same time. I'm guessing this has to do with the parameters being shared, I noticed that some moved the same parameter independent from the realearn fx.
But if they share the parameters I'm a bit confused as to why changing a mapping group for another realearn instance doesn't work.

Edit: I had one controller set as fx that is why the parameters were shared. Is it maybe possible to change mapping groups when control method is set to fx? But so far I couldn't get multiple controllers work well together when using fx control method.
The issue isn't gone completely but it could be my controller that is bust. I'll have to test is some more.

**Reply by helgoboss** [reply]
You can use the ReaLearn instance with input "MPD controller" to switch the banks of the ReaLearn instance with input "Studiologic controller". In the target "FX parameter: Set value", simply don't use FX "This" but select the "MPD controller" ReaLearn FX.

About your second question: So you want different mappings depending on which FX is focused? Then follow the "Auto-load" tutorial in the user guide.

**Comment by BrahRah** [comment]
I looked through all of them and there is not an option to select the controller directly. Or do you mean to select the fx that uses the controller?
![no such option](https://user-images.githubusercontent.com/40721994/197341501-6c98f10c-e151-44b7-804f-a0aab11c90c0.png)



I've also tested it further and the 100 parameter limit seems to create parameters that are shared between the controllers which messes up using multiple controllers for one vst. Is it possible to get more parameters?

**Reply by helgoboss** [reply]
Mmh, if you perceive the 100 parameters as limit, then you probably use ReaLearn in a way which is not intended. Why the need for so many parameters? Can you send me an export or the complete project so I know what I'm dealing with here? Along with some comments.

Two things:

- It's best practice (and almost required) to use at least one ReaLearn instance per controller. Then a mixup between multiple controllers can't even happen. And you have 100 parameters per compartment per instance, that should be more than enough. And you can always fire up a second instance for the same controller, no problem.
- As for the target question: Simply press "Learn target", move the parameter slider of the other ReaLearn instance whose parameter you want to control (UI button reveals the list of parameters of the instance) and that's it. Alternatively, you can do that manually by choosing the track, usually using "By ID" and then choosing the FX, also usually "By ID" and then the parameter. If you don't know what all this means, have a look into the user guide, search for it.

**Reply by helgoboss** [reply]
And just to make this clear: ReaLearn parameters are only necessary for conditional activation (1 parameter = 1 modifier, 1 parameter = many banks) or dynamic selection of track/FX and other things. 

It has nothing to do with controlling parameters of other plugins.

**Comment by BrahRah** [comment]
I didn't save the project and after recreating it the issue was not there anymore. I'm assuming I messed up somewhere. 
I've got one instance per controller but all 3 ReaLearn instances on the same track.

**Reply by 7enz** [reply]
i believe this does work, 
however it works mainly with endless encoders
provided feedback is set and the same midicc's & channels are used across both units, it can be used

with faders its hit and miss.. depending on if they are motorised and if they can receive and send feedback
i have this type of use case setup across 2 maschine jam units& a BCR2000... where the led touchstrips provide feedback for both 2nd maschine jam unit and the BCR2000 or vice versa
